README.md
# qhttotxt - QIP history (.qhf, .ahf and .txt) to .txt combiner and converter #

### Prehistory ###
At the end of the 00s and at the beginning of the 10s QIP messenger got widespread. It worked on ICQ protocol. The 
messenger had a mobile version - QIP PDA. As a difference from other messengers storing their history in text files, 
QIP stored it in internal formats: .qhf and .ahf. Moreover, QIP for PC encrypted the history as well. Besides, as most 
messengers at that time, there was no synchronization between devices and it was possible when messages from the same 
contact were in different places. And if there is a folder "Archive" with .ahf files, messages from the same contact 
will be in separate files as well. Finally, the history can be stored in plain txt files form messengers like Mchat or 
converted from ICQ, QIP and RnQ with third-party tools. This makes a lot of mess and combining of files takes a lot of 
efforts.

Several programs have been developed to translate QIP history into text files, however, as of April 2021, they are 
either not available or do not work on modern OSs. Anyway, the author was not able to find a working version in the 
public domain.

### Short description ###

qhttotxt is a small Java program that allows converting QIP internal history files (.qhf and .ahf) into human-readble 
.txt files. The program is able to combine messages from different files for one contact, choosing unique ones and 
putting them in accordance with the time of sending/receiving. In version 1.0a it is possible to get history from txt 
files with certain formats (see below), and combine and save contact lists from .cl and .cbd files.

Also, there are minimal options for recovering corrupted messages and files. However, the author does not have a large 
number of files for testing, so it is likely that some corrupted files will contain missing messages.

### Usage ### 
JAVA JDK version higher than 1.8 is needed to be installed. 

The simplest way:
Download /jars/qhftotxt.jar from this repository and put it into a folder with history files. Run the program without 
any keys:

`> java -jar qhttotxt.jar`

The program will find all possible files in the given folder and convert them.

Advanced options are available using parameters (can be in any order):

`{path}` - the path of the directory with files or the path to a single file. Only one path is supported. If you need 
to convert several folders, place them into one root folder. Use the `-r` switch and specify the path to root folder.

`-с` - combining messages from one contact. If the `-r` switch is provided, then the output files will be placed in 
the working directory. Only unique messages are combined, all duplicates are removed. When the key `-c` is set, there 
will be an attempt to find txt files in QIP/ICQ, Mchat and RnQ formats (formats description is below, as well as an 
**important** requirements to thier names). There will be an attempt to find contact lists files .cl and .cbd, combine 
and save them to a single one. Only UINs, nicknames and groups are used from contact lists. 

`-h` - short help.

`-n {nickname}` - there is no information about the owner's nickname or UIN in the history files, so it is  
impossible to restore it. If the nickname is not specified by this key, then it will be 'You' in all texts by default.

`-p {codepage}` - force setting of a codepage. If the program runs on a computer that does not have a language of 
history files, this key may help. The codepage must be specified in the standard format. See details 
https://en.wikipedia.org/wiki/Code_page.

`-r` - recursive search in subdirectories of the working directory. By default, recursive search is disabled. The 
output files will be placed in the same directories as their sources.

`-z {timezone}` - force setting of timezone for displaying message time. If the time zone is different on a computer 
where the program runs from the one where the history was recorded, then the time in output files will be displayed 
incorrectly. The time zone must be specified in a standard format. See details 
https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

Examples:
Convert files in a current (the one with file `qhttotxt.jar`) Directory:

`> java -jar qhttotxt.jar`

Convert file `C:\Users\User\Documents\file.qhf` only (file `qhttotxt.jar` can be in any place):

`> java -jar qhttotxt.jar C:\Users\User\Documents\file.qhf`

Covert files in `C:\Users\User\Documents` and all subdirectories, set the nickname "John Snow" and codepage 
"windows-1251" (standard for Russian). File `qhttotxt.jar` can be in any place:

`> java -jar qhttotxt.jar -r -n "John Snow" -p "windows-1251" C:\Users\User\Documents`

Covert files in `C:\Users\User\Documents` and all subdirectories, combine histories from different files, set the 
nickname "John Snow" and codepage "windows-1251" (standard for Russian), timezone "Europe/Moscow" (full set of keys). 
File `qhttotxt.jar` can be in any place: 

`> java -jar qhttotxt.jar -c -n "John Snow" -p "windows-1251" -z "Europe/Moscow" -r C:\Users\User\Documents`

## Description of text file formats recognizing by the program ##

If your files look like the formats below, qhftotxt will try to read the and combine with other existing files. 
**Important!** Filenames of txt files **MUST** be `{contact_uin}.txt`, i.e. `123456789.txt`. It is the name of the file 
that will determine the UIN of the contact. If the filename has a different pattern, change it manually. If there are 
some files with the same name, just put them into subdirectories and use `-r` key.

### qip/icq: ###
``` 
{Beginning of the file}
-------------------------------------->- (separator, `<` - received message, `>` - sent message)
{Owner_nick} (21:48:51 18/08/2015)
message


--------------------------------------<-
{contact_nick} (14:28:50 19/10/2015)
Message


--------------------------------------<-
... 
```

### Mchat: ###
```
{Beginning of the file}
28.05.08 10:13:17<message
28.05.08 10:15:26>message
28.05.08 10:21:39<message
message str2
28.05.08 10:21:59>message
```
### RnQ (converted with RnQHistoryReader) ###
```
{Beginning of the file}
Чат между {Owner_uin} и {contactUin}


13.10.2006 10:48:00 {contactUin}
message


13.10.2006 10:48:40 {Owner_uin}
message


13.10.2006 10:54:58 {contactUin}
message


```

## QHF (AHF) format description ##

Original format description was taken from https://github.com/MolinRE/QIParser. Without this description, this program 
would not exist. Many thanks to the author and owner Konstantin Komarov (https://github.com/MolinRE) for permission 
for using it in this project as well. There is another description 
https://alexey-m.ru/articles/qip-infium-prodolzhenie-istorii 

## File structure ##

### File header ###

Header includes the history file info. From there, you can find out the size of the story (it is not clear why), the 
number of messages (it is not clear why it goes twice), the UIN and the nickname of the person. All numeric values 
are stored as `Big Endian`. Strings are stored in `UTF8`.

Position | HEX | Size | Description
------------- | ------------- | ------------- | -------------
0 | 0x00 | Char(3) | Signature. Always "QHF".
4 | 0x04 | Int32 | History size in bytes.
34 | 0x22 | Int32 | Number of messages.
38 | 0x26 | Int32 | ???
44 | 0x2C | Int16 | UIN length.
46 | 0x2E | Char(N) | UIN. N = length.
46 + N | 0x2E + N | Int16 | Nickname length.
48 + N | 0x30 + N | Char(N) | Nickname. N = length.

### Messages ###
Messages are right after the header. They are going one by one. The position is counted after the heading.

_Comment of serg987: analysis of ~ 50K messages shows that the values of most of the fields are always the 
same (see the "Value" column)_ 

Position | HEX | Size | Description | _Value_
------------- | ------------- | ------------- | -------------  | -------------
00 | 0x00 | Int16 | Signature. Always 1. | 1
02 | 0x02 | Int32 | Size of the message block. (w/o first 6 bytes)
06 | 0x06 | Int16 | Type of (?) message ID field. | 1
08 | 0x08 | Int16 | Size of the block with message ID. | 4
10 | 0x0A | Int32 | Message ID. Just a serial number. 
14 | 0x0E | Int16 | Type of (?) message sending date field. | 2
16 | 0x10 | Int16 | Size of message sending date field. | 4
18 | 0x12 | Int32 | Date UNIX format.
22 | 0x16 | Int16 | Type of field ??? | 3
24 | 0x18 | Int16 | ??? | 3
26 | 0x1A | Byte | Is the message sent. Boolean.
27 | 0x1B | Int16 | Type of message text field. | see table below
29 | 0x1D | Int16 | Size of the field of message length value. _(1)_ | 4
31 | 0x1F | Int16/32 _(1)_ | Message length in bytes.
35 | 0x23 | Byte(N) | Message text, _(non)_ encrypted. _(2)_

_Comment of serg987:_

_(1) - For QIP PDA files with non-encrypted message text Int16 (2 bytes) are used for length, but the value of the field 
0x1D - still equals 04. It may be a bug or maybe it is not a field size, but just another constant._

_(2) - For QIP PDA files message text is not encrypted._

### Message encryption ###
_Comment of serg987: not applicable for QIP PDA_

The text of the messages is stored in UTF8 encoding, as mentioned earlier. However, before converting the bytes to 
text, they are needed to be decrypted. The decrypting of the bytes of the message text is performed by the following 
principle:
```csharp
for (int i = 0; i < array.Length; i++)
    array[i] = (byte)(255 - array[i] - i - 1);
```

### Detailed field description ###
Each block is preceded by a block identifier of a 16-bit signed integer (or it is unsigned?). Here is a table below 
for the possible types of blocks with some details.

_Comment of serg987: according to observations, fields of type 05, 13, and 14 are duplicated by a corresponding phrase 
in the message text_

Type | Value
------------- | -------------
01 | Online message.
02 | Message sending date.
03 | Message sender.
05 | Authorization request.
06 | Friend request.
13 | Offline message.
14 | Authorization request accepted.
| | Added by serg987:
80 | QIP/ICQ service messages. Usually for connection issues.
81 | QIP/ICQ service messages. Usually for contacts' birthdays.

### License ###

Standard MIT license
Copyright (c) 2021 Sergey Kiselev

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
Changelog:
ver 1.0a:
  - Added parsing and combining for QIP PDA (.cl) and Mchat (.cbd) files (works with -c key only). Combined file is saved
in working directory, and the data from contact lists is used for creating chat filenames and chat header.
  - Added parsing and combining of .txt files for Mchat, QIP-ICQ (exported?) and RNQ (exported with RnQHistoryReader.exe)
chats. Messages taken from txt files are combined with QHF and AHF so the whole history can be saved in a single file.
  - Charset of files is determined before storing it for better consistency. But it is still needed to pass it in 
commandline for machines without language of input files.
  - Extra logging.
  - As the author achieved all the aims with this functionality, no other developing is planned unless there are bugs
or really important features to implement. 

ver 0.9a:
- First public release
- Parsing and combining QHF and AHF files

src/main/java/RegexScratch.java
import java.time.ZoneId;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RegexScratch {

    public static void main(String[] args) {
        String str1 = "-------------------------------------->-";
        String str2 = "--------------------------------------<-";
        String str3 = "-------------------------------------<-";

        String pattern1 = "^[-]{38}[<>][-]";
        System.out.println(str1.matches(pattern1));
        System.out.println(str2.matches(pattern1));
        System.out.println(str3.matches(pattern1));

        str1 = "Kisser (19:32:35 20/04/2015)";
        str2 = "D E (19:31:47 20/04/2015)";
        str3 = "Антонио (19:29:17 20/04/2015)";

        pattern1 = "^[\\d|\\p{L}|\\s|@|\\.]*\\s[(][\\d|:]*\\s[\\d|\\/]*[)]";
        System.out.println(str1.matches(pattern1));
        System.out.println(str2.matches(pattern1));
        System.out.println(str3.matches(pattern1));

        str1 = "09.12.08 18:03:44>тут еще?";
        str2 = "11.12.08 10:21:00<надо будет как-нибудь взять  ;-)";
        str3 = "11.12.08 10:2!:00<надо будет как-нибудь взять  ;-)";
        Pattern mchatLineHeaderPattern = Pattern.compile("^([\\d]{2}[\\/|\\.]){2}[\\d]{2,4}\\s[\\d]{1,2}([:][\\d]{2}){2}[<|>]");
        Matcher m = mchatLineHeaderPattern.matcher(str1);
        if (m.find()) {
            System.out.println(m.group());
        } else System.out.println("No matches");
        m = mchatLineHeaderPattern.matcher(str2);
        if (m.find()) {
            System.out.println(m.group());
        } else System.out.println("No matches");
        m = mchatLineHeaderPattern.matcher(str3);
        if (m.find()) {
            System.out.println(m.group());
        } else System.out.println("No matches");

        System.out.println(str3.replace(".txt", ""));

        str1 = "09.12.08 09:03:44";
        Configuration.zoneId = ZoneId.of("Europe/Moscow");
        System.out.println(Commons.parseDateTime(str1));

        String str4 = "aaa";
        str4 = str4.concat("ddd").concat("bbb");
        System.out.println(str4);

    }
}
